### PR TITLE
Fix #687: Refactor store initialization

### LINKED ISSUE #687: Refactor GLAM's store instantiation to work like GUD's
Part of the Fenix story in #569

The store in GLAM has a lot of hard-coded functionality that gets in the way of generalizing for other products. One critical issue is the store instantiation on page load.

GUD has a nice way of instantiating the store that doesn't quite hard-code as much stuff. We could easily adapt GLAM's approach to follow that approach.

### PR DESCRIPTION
**Pull request by openjck**
Wanted to share my progress on this. This is untested, but I think it's a step in the right direction.

I'd like to merge some aspects of *firefox-desktop.js* into *shared.js* but may leave that for a later PR.

### CODE REVIEW

File: src/config/shared.js [hunk 1]
```diff
@@ -32,3 +32,38 @@ export function extractBucketMetadata(transformedData) {
...
+    defaultValue: 'nightly',
+  },
+  os: {
+    defaultValue: 'Windows',
+  },
+  process: {
+    defaultValue: 'content',
+  },
+  product: {
+    defaultValue: 'firefoxDesktop',
```
Comment (robhudson) [inline src/config/shared.js]: Just a caveat... the Glean name for Firefox will be `firefox.desktop`, if I saw that correctly. The API currently expects just `firefox`. I'm happy to change it to something that is more clearly "telemetry firefox" vs Glean firefox.

File: src/config/shared.js [hunk 2]
```diff
@@ -32,3 +32,38 @@ export function extractBucketMetadata(transformedData) {
...
+    defaultValue: [],
+  },
+  aggregationLevel: {
+    defaultValue: 'build_id',
+  },
+  channel: {
+    defaultValue: 'nightly',
+  },
+  os: {
+    defaultValue: 'Windows',
```
Comment (robhudson) [inline src/config/shared.js]: I've always felt the default here should be `*` for all OS's. And if the user wants to drill down they can. What do others think?

File: src/config/shared.js [hunk 3]
```diff
@@ -32,3 +32,38 @@ export function extractBucketMetadata(transformedData) {
...
+    defaultValue: '',
+  },
+  timeHorizon: {
+    defaultValue: 'MONTH',
+  },
+  visiblePercentiles: {
+    isMulti: true,
+    defaultValue: [95, 75, 50, 25, 5],
+  },
+};
```
Comment (robhudson) [inline src/config/shared.js:51]: Another item to possibly add here, or maybe it comes with a future code change, is `versions`, as in, the number of versions to query data for, defaulting to 3 if not sent to the API.

**Comment by hamilton** [comment]
yeah, let's leave the moving of stuff from firefox desktop to shared for another PR.

File: src/config/shared.js [hunk 4]
```diff
@@ -32,3 +32,38 @@ export function extractBucketMetadata(transformedData) {
   etc.bucketSortOrder = sorter;
   return etc;
 }
+
+export const facets = {
```
Comment (hamilton) [inline src/config/shared.js]: Let's wait to do this for another PR. The default value for channels in other products are (for instance) not nightly. They're usually release. It'll be clearer what belongs in shared when we get our bearings, which will probably happen once people can see the data in GLAM.

File: src/config/firefox-desktop.js [hunk 5]
```diff
@@ -12,47 +12,67 @@ import { validate, noResponse } from '../utils/data-validation';
...
-      title: 'Channel',
-      key: 'channel',
-      values: [
-        { key: 'nightly', label: 'Nightly' },
-        { key: 'beta', label: 'Beta' },
-        { key: 'release', label: 'Release' },
-      ],
-      isValidKey(key, probe) {
-        return (probe.prerelease && key !== 'release') || !probe.prerelease;
+    productDimensions: {
```
Comment (hamilton) [inline src/config/firefox-desktop.js]: Can you give more details why we have `dimensions.productDimensions` and not just `dimensions` here? This seems like an unnecessary nesting

File: src/state/store.js [hunk 6]
```diff
@@ -5,21 +5,29 @@ import { createStore } from '../utils/create-store';
...
+  return pathProduct ? mappings[pathProduct] : defaultProduct;
+}
 
-// TODO: move this to the new config.js when 'product' is added.
-const DEFAULT_PROBE_PROCESS = 'content';
+const CONFIG = productConfig[getProduct()];
 
 export function getField(fieldKey) {
-  return CONFIG.dimensions[fieldKey];
+  return CONFIG.dimensions.productDimensions[fieldKey];
```
Comment (hamilton) [inline src/state/store.js]: These `getField` type functions are hard-coded to `CONFIG`, which means that whatever product is set on page load is what `CONFIG` resorts to. It seems to me that we should be using `getProduct` in each of these functions where appropriate. It's unclear how this will work with a different product as-is.

**Review by hamilton — CHANGES_REQUESTED** [review]
This is looking good overall. Had two questions:

1. can you give more context for having `dimensions.productDimensions` vs. just `dimensions`? It's unclear what the additional nesting accomplishes.
2. there is more work to be done here re: initialization and how it will play with future products. Your `getProduct` function is a good start, but `CONFIG` is hard-coded to whatever product was there on page load, not future ones. Every single place where `CONFIG` is called should probably keep in mind what the _current_ product is.

File: src/state/store.js [hunk 7]
```diff
@@ -5,21 +5,29 @@ import { createStore } from '../utils/create-store';
 // FIXME: take care of this dependency cycle.
 import { probeSet } from './telemetry-search'; // eslint-disable-line import/no-cycle
 
-import CONFIG from '../config/firefox-desktop';
+import productConfig from '../config/products';
+
+function getProduct() {
+  const defaultProduct = 'firefoxDesktop';
+  const mappings = {
```
Comment (hamilton) [inline src/state/store.js]: this mapping should probably live in `../config/products`, not here. That's the point of the config file.

Comment (hamilton) [reply src/state/store.js]: Also, we should remove the _need_ for this mapping. Are we not coding things properly b/t the API & GLAM? This should be a goal imo. Might be worth getting @robhudson's input here. If we have to ahve this mapping, that's fine, but would be worth filing an issue to fix this sooner rather than later so we don't have to maintain this.

**Review by hamilton — CHANGES_REQUESTED** [review]
left out a few additional points

File: src/state/store.js (hunk 7 above)
Comment (hamilton) [reply src/state/store.js]: Furthermore, we could have `getProduct` be a function in `productConfig` I think.

File: src/config/shared.js (hunk 2 above)
Comment (openjck) [reply src/config/shared.js]: >     1. can you give more context for having `dimensions.productDimensions` vs. just `dimensions`? It's unclear what the additional nesting accomplishes.

I'm fine removing that. It would make things easier. But I thought we agreed on Slack that we didn't want to flatten it out.

**Comment by openjck** [comment]
> 1. can you give more context for having `dimensions.productDimensions` vs. just `dimensions`? It's unclear what the additional nesting accomplishes.

I'm fine with removing the `productDimensions` object. It would make things easier. But I thought we agreed on Slack that we didn't want to flatten it out.

> 2. there is more work to be done here re: initialization and how it will play with future products. Your `getProduct` function is a good start, but `CONFIG` is hard-coded to whatever product was there on page load, not future ones. Every single place where `CONFIG` is called should probably keep in mind what the _current_ product is.

I see. Yes, that's true. I'll look at that.

File: src/state/store.js [hunk 8]
```diff
@@ -1,25 +1,91 @@
...
+    auth: store.getState().auth,
+    probeName: '',
+  });
+};
+
+const active = {
+  get productConfig() {
+    return productConfig[get(store).product];
+  },
+};
```
Comment (openjck) [inline src/state/store.js]: This allows us to do `active.productConfig.dimensions.[whatever]`. Another pattern I considered was adding a `get` method to the imported `productConfig` object, which would allow us to do `productConfig.currentProduct.dimensions.[whatever]`, which would look like this:

```javascript
Object.defineProperty(productConfig, 'currentProduct', {
  get() {
    return productConfig[get(store).product];
  },
});

const whatever = productConfig.currentProduct.dimensions.[whatever];
```

I'd be interested to hear which you think is more readable.

Comment (hamilton) [reply src/state/store.js]: I think the latter is a bit more readable to me b/c it's tied a bit more obviously to the config so I know where to look for the value if I see `productConfig.currentProduct` elsewhere. Also aren't we going to need `productConfig.currentProduct` elsewhere, such as in the dimension selector region? Is the goal to have other potential active things in `active`? If so, then I think this is a good approach, but I can't think of another case that we'd need this. My pref would be to go for `productConfig.currentProduct` with this in mind, but this is a weak position – either works for me.

File: src/config/shared.js (hunk 3 above)
Comment (hamilton) [reply src/config/shared.js:51]: Let's make sure to file these as follow-on issues.

File: src/state/store.js [hunk 9]
```diff
@@ -1,25 +1,91 @@
...
+      } else {
+        state.productDimensions[key] = defaultValue;
+      }
+    }
+  );
+
+  return state;
+}
+
+export const store = createStore(getDefaultState());
```
Comment (hamilton) [inline src/state/store.js]: should this have `basedOnQueryParams: true`?

File: src/config/shared.js [hunk 10]
```diff
@@ -32,3 +32,20 @@ export function extractBucketMetadata(transformedData) {
   etc.bucketSortOrder = sorter;
   return etc;
 }
+
+export default {
+  activeBuckets: {
```
Comment (hamilton) [inline src/config/shared.js:37]: shouldn't `reference` also be here in some way? It doesn't look like the reference point persists now.

**Review by hamilton — CHANGES_REQUESTED** [review]
This is looking great. I left a few comments about other things I'm noticing as I test out these changes:

1. some parameters don't seem to persist. It seems like that's probably due to `basedOnQueryParams` being `false` by default. Easy fix.
2. even if I set `basedOnQueryParams` manually on my local copy, it doesn't appear that the reference point persists from pageload to pageload. Probably worth taking a look at that, & I'm sure it's an easy fix.
3. there's a potential merge conflict with the changelog that'll need to be fixed before merging.

Other than these parts, I think this looks good. Feel free to address those and test / merge at your discretion. Thanks!

File: src/state/store.js (hunk 8 above)
Comment (openjck) [reply src/state/store.js]: Sounds good. We may need `currentProduct` elsewhere, but I think *store.js* is mainly where things get weird: we need to read from the store itself to determine what `currentProduct` is.

I could add a `currentProduct` property to *config/products.js*. That would create a dependency cycle. I'm not certain that the dependency cycle would cause problems, but I try to avoid them in general.

File: src/config/shared.js (hunk 3 above)
Comment (openjck) [reply src/config/shared.js:51]: @robhudson Would you mind opening issues for this and the thing about firefox.desktop? I don't completely understand what you mean. I'll open an issue about OS defaults.

File: src/config/firefox-desktop.js (hunk 5 above)
Comment (openjck) [reply src/config/firefox-desktop.js]: We hashed this out on Slack

File: src/state/store.js (hunk 8 above)
Comment (openjck) [reply src/state/store.js]: I opted for something a little easier with less magic. May have been overthinking it.

File: src/state/store.js (hunk 9 above)
Comment (openjck) [reply src/state/store.js]: Yes, important catch. Thank you!

File: src/config/shared.js (hunk 1 above)
Comment (openjck) [reply src/config/shared.js]: Would you mind opening a follow-up issue for this?

File: src/config/shared.js (hunk 2 above)
Comment (openjck) [reply src/config/shared.js]: I opened an issue for this: #753
